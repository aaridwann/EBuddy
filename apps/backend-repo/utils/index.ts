import { DataOrError } from './utils.types';

/**
 * Generate Response
 * @param success 
 * @param dataOrError 
 * @returns 
 */
export const generateResponse = (success: boolean, dataOrError: DataOrError) => {
  if (success) {
    return {
      success: true,
      data: dataOrError,
    };
  }

  return {
    success: false,
    error: {
      code: dataOrError.code || 500,
      message: dataOrError.message || 'Unknown error',
    },
  };
};
