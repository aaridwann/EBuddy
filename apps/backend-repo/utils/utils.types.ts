export type Error = { code: number, message: string };
export type Data = { [key: string]: string | boolean | number };

export type DataOrError = Data | Error;