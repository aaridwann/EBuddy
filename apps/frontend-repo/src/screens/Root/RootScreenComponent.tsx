'use client';

import React, { useEffect, useState } from 'react';
import { onAuthStateChanged } from 'firebase/auth';
import { useRouter } from 'next/navigation';
import { get, isNull, mapValues, toPairs } from 'lodash';
import { Avatar, Box, Button, Card, CardContent, Container, Divider, Skeleton, Typography } from '@mui/material';

import UpdateButton from '@/components/Button/Button';
import i18n from '@/localize';
import LoadingComponent from '@/components/Loading';

import { auth } from '../../../../../packages/shared/src/firebase/firebaseConfigs';

import { HandleSubmit, HooksType, Props } from './RootScreen.types';
import { defaultAlt, defaultImage, initialData, shimmerMapContent } from './RootScreen.configs';
import EditForm from './EditForm/EditFormComponent';
import Styles from './RootScreenComponent.styles';

/**
 * AuthChanged hooks
 * @returns {HooksType} - Hooks
 */
const Hooks = (props: Props): HooksType => {
  const router = useRouter();
  const [token, setToken] = useState<string | null>(null);
  const [formData, setFormData] = useState(initialData);
  const [isEdit, setIsEdit] = useState(false);

  const handleChange = React.useCallback((key: string, value: string) => {
    setFormData((prev) => ({ ...prev, [key]: value }));
  }, []);
  const isEditHandler = React.useCallback(() => setIsEdit((prev) => !prev), []);

  useEffect(() => {
    onAuthStateChanged(auth, async (user) => {
      if (user) {
        const idToken = await user.getIdToken();
        setToken(idToken);
      } else {
        router.replace('/login');
      }
    });
  }, [router]); 

  useEffect(() => {
    const data = mapValues(get(props, 'user.data'), (value) => isNull(value) ? '' : value);

    if (data) {
      setFormData(data);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.user.data]);

  return { token, handleChange, formData, isEditHandler, isEdit, router };
};

/**
 * Update handler
 * @param {Props} props - Props
 * @param {string | null} token - Token
 * @returns {VoidFunction} - Update handler
 */
export const _updateHandler = (props: Props, token: string | null): VoidFunction => () => {
  if (isNull(token)) {return;}

  props.fetchUser();
};

/**
 * Logout handler
 * @param {Props} props - Props
 * @param {HooksType} hooks - Hooks
 * @returns {VoidFunction} - Logout handler
 */
const _logoutHandler = (props: Props, hooks: HooksType): VoidFunction =>  () => {
  props.logoutUser();
  hooks.router.replace('/login');
};

/**
 * Success handler
 * @param {HooksType} hooks - Hooks
 * @returns {VoidFunction} - Success handler
 */
const _successHandler = (hooks: HooksType): VoidFunction => async () => {
  hooks.isEditHandler();
};

/**
 * Handle submit
 * @param {Props} props - Props
 * @param {HooksType} hooks - Hooks
 * @returns {VoidFunction} - Handle submit
 */
const _handleSubmit: HandleSubmit = (props: Props, hooks: HooksType) => (e: React.ChangeEvent) => {
  e.preventDefault();
  props.updateUser(hooks.formData)
    .then(_successHandler(hooks));
};

/**
 * Render avatar
 * @param {Props} props - Props
 * @returns {React.ReactNode} - Render avatar
 */
const _renderAvatar = (props: Props): React.ReactNode => {
  const { auth: { user } } = props;
  const photoURL = get(user, 'photoURL');
  const displayName = get(user, 'displayName');

  if (!user) {return <Skeleton sx={Styles.skeleton} variant='circular' width={100} height={100} />;}

  return (
    <Avatar
      src={photoURL || defaultImage}
      alt={displayName || defaultAlt}
      sx={Styles.avatar}
    /> 
  );
};

/**
 * Render shimmering profile
 * @returns {React.ReactNode} - Render shimmering profile
 */
const _renderShimmeringProfile = (): React.ReactNode => (
  <React.Fragment>
    {shimmerMapContent.map(({ width, height }, index: number) => (
      <Skeleton key={index} sx={Styles.skeletonProfile} variant='rounded' width={width} height={height} />
    ))}
  </React.Fragment>
);

/**
 * Render card
 * @param {Props} props - Props
 * @returns {React.ReactNode} - Render card
 */
const _renderCard = (props: Props, hooks: HooksType): React.ReactNode => {
  const { auth: { user } } = props;
  const email = get(user, 'email');
  const displayName = get(user, 'displayName');

  if (!user) {return _renderShimmeringProfile();}

  return (
    <CardContent>
      <Typography variant="h5" gutterBottom>{displayName || 'Anonymous'}</Typography>
      <Typography variant="body1" color="textSecondary">{email}</Typography>
      <Box mt={2}><Button onClick={_logoutHandler(props, hooks)} variant="contained" color="primary">{i18n.t('LOGOUT')}</Button></Box>
    </CardContent>
  );
};

/**
 * Render Button Details
 * @param {HooksType} hooks - Hooks
 * @param {Props} props - Props
 * @returns {React.ReactNode} - Render Button Details
 */
const _renderButtonDetails = (hooks: HooksType, props: Props): React.ReactNode => {
  const title = !isNull(props.user.data) ? 'EDIT_USER_DATA': 'FETCH_USER_DATA';
  const { token, isEditHandler } = hooks;

  return (
    token && <UpdateButton 
      title={i18n.t(title)}
      loading={props.user.loading} 
      error={props.user.error} 
      onClick={isNull(props.user.data) ? _updateHandler(props, token) : isEditHandler} 
    />
  );
};

/**
 * Render details result
 * @param {Props} props - Props
 * @returns {React.ReactNode} - Render details result
 */
const _renderDetails = (props: Props): React.ReactNode => {
  const data = get(props, 'user.data', []);
  const transformData = toPairs(data);

  return (
    transformData.map(([keyBy, value]) => (
      <React.Fragment key={keyBy}>
        <Box maxWidth='sm' sx={Styles.box}>
          <Typography variant="h5" gutterBottom>{keyBy}</Typography>
          {keyBy === 'photoURL' ? 
            <Avatar
              src={value || '/default-avatar.png'}
              alt={value || 'User'}
              sx={Styles.avatarDetails}
            />  
            : 
            <Typography variant="h6" gutterBottom>{value}</Typography>}
        </Box>
      </React.Fragment>
    ))
  );
};

/**
 * Render details content
 * @param {HooksType} hooks - Hooks
 * @param {Props} props - Props
 * @returns {React.ReactNode} - Render details content
 */
const _renderDetailsContent = (hooks: HooksType, props: Props): React.ReactNode => (
  <Box maxWidth='sm' sx={Styles.boxDetails}>
    {_renderDetails(props)}
    {_renderButtonDetails(hooks, props)}
  </Box>
);

/**
 * Render edit form and details
 * @param {Props} props - Props
 * @param {HooksType} hooks - Hooks
 * @returns {React.ReactNode} - 
 */
const _renderEditDetails = (props: Props, hooks: HooksType): React.ReactNode => (
  hooks.isEdit ? 
    <EditForm props={props} hooks={hooks} handleSubmit={_handleSubmit} /> 
    : _renderDetailsContent(hooks, props)
);

/**
 * Render content
 * @param {Props} props - Props
 * @param {HooksType} hooks - Hooks
 * @returns {React.ReactNode} - Render content
 */
const _renderContent = (props: Props, hooks: HooksType): React.ReactNode => (
  <React.Fragment>
    <Container maxWidth="xl" sx={Styles.container} >
      <Card sx={Styles.cardContent}>
        {_renderAvatar(props)}
        {_renderCard(props, hooks)}
      </Card>
    </Container>
    <Divider sx={Styles.divider}>Details</Divider>
    {_renderEditDetails(props, hooks)}
  </React.Fragment>
);

/**
 * Root Screen Component
 * @param {Props} props - Props
 * @returns {React.ReactNode} - Root Screen Component
 */
const RootScreenComponent = (props: Props): React.ReactNode => {
  const hooks = Hooks(props);
  const { auth } = props;

  if (!auth || isNull(hooks.token)) {return <LoadingComponent size={45}  />;}

  return _renderContent(props, hooks);
};

export default RootScreenComponent;
