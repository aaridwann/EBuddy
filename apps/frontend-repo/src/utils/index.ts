import { AxiosError } from 'axios';
import { get } from 'lodash';

const getErrorCode = (err: Partial<AxiosError>) => {
  const errorCode = get(err, 'payload.response.data.error.code', 'INTERNAL_ERROR');
  
  return errorCode;
};

export { getErrorCode };