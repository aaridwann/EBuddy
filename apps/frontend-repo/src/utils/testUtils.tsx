import * as React from 'react';
import { render } from '@testing-library/react';

interface Configs {
  props: {[key: string]: unknown},
  desc: string
}

export const assertSnapshot = (configs: Configs[], Component: React.ElementType) => {
  configs.map(({ props, desc }) => (
    it(desc, () => {
      expect(render(<Component {...props} />)).toMatchSnapshot();
    })
  ));
};