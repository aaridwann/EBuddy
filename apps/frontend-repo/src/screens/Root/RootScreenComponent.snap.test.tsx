/* eslint-disable max-lines-per-function */
import { assertSnapshot } from '@/utils/testUtils';

import RootScreenComponent from './RootScreenComponent';

jest.mock('next/navigation', () => ({
  __esModule: true,
  ...jest.requireActual('next/navigation'),
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn(),
    replace: jest.fn(),
  }),
}));

describe('RootScreenComponent snapshot testing', () => {
  const configs = [
    {
      props: {
        user: {
          data: null,
          loading: false,
          error: null,
        },
        auth: {
          user: null,
          token: null,
          loading: false,
          error: null,
          displayName: null,
        },
        fetchUser: jest.fn(),
        updateUser: jest.fn(),
        logoutUser: jest.fn(),
      },
      desc: 'should render when value of props is null',
    },
    {
      props: {
        user: {},
        auth: {
          user: null,
          token: null,
          loading: false,
          error: null,
          displayName: null,
        },
        fetchUser: jest.fn(),
        updateUser: jest.fn(),
        logoutUser: jest.fn(),
      },
      desc: 'should render when user is null',
    },
    {
      props: {
        user: {
          id: '123456',
          uid: '123456',
          displayName: 'John Doe',
          photoURL: 'https://pinterest.com/image.png',
          email: 'johndoe@mail.com',
          token: 'abcdefg123456',
          loading: false,
          error: null,
        },
        auth: {
          user: {
            uid: '123456',
            displayName: 'John Doe',
            photoURL: 'https://pinterest.com/image.png',
            email: 'johndoe@mail.com',
          },
          token: 'abcdefg123456',
          loading: false,
          error: null,
          displayName: 'John Doe',
        },
        fetchUser: jest.fn(),
        updateUser: jest.fn(),
        logoutUser: jest.fn(),
      },
      desc: 'should render when value of props is not null',
    },
  ];

  assertSnapshot(configs, RootScreenComponent);
});