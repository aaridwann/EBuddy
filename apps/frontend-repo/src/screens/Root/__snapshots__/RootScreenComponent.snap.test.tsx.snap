// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootScreenComponent snapshot testing should render when user is null 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <p
      class="firebase-emulator-warning"
      style="position: fixed; width: 100%; background-color: rgb(255, 255, 255); border: .1em solid #000000; color: rgb(181, 0, 0); bottom: 0px; left: 0px; margin: 0px; z-index: 10000; text-align: center;"
    />
    <div>
      <div
        class="MuiBox-root css-1lb6n3v"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
          role="progressbar"
          style="width: 45px; height: 45px;"
        >
          <svg
            class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1lb6n3v"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
        role="progressbar"
        style="width: 45px; height: 45px;"
      >
        <svg
          class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RootScreenComponent snapshot testing should render when value of props is not null 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <p
      class="firebase-emulator-warning"
      style="position: fixed; width: 100%; background-color: rgb(255, 255, 255); border: .1em solid #000000; color: rgb(181, 0, 0); bottom: 0px; left: 0px; margin: 0px; z-index: 10000; text-align: center;"
    />
    <div>
      <div
        class="MuiBox-root css-1lb6n3v"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
          role="progressbar"
          style="width: 45px; height: 45px;"
        >
          <svg
            class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1lb6n3v"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
        role="progressbar"
        style="width: 45px; height: 45px;"
      >
        <svg
          class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`RootScreenComponent snapshot testing should render when value of props is null 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <p
      class="firebase-emulator-warning"
      style="position: fixed; width: 100%; background-color: rgb(255, 255, 255); border: .1em solid #000000; color: rgb(181, 0, 0); bottom: 0px; left: 0px; margin: 0px; z-index: 10000; text-align: center;"
    />
    <div>
      <div
        class="MuiBox-root css-1lb6n3v"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
          role="progressbar"
          style="width: 45px; height: 45px;"
        >
          <svg
            class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1lb6n3v"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
        role="progressbar"
        style="width: 45px; height: 45px;"
      >
        <svg
          class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
